File: FirstNonRepeatingCharacter.java

import java.util.Scanner;
import java.util.HashMap;
import java.util.Map; 

class Index { 
  int count;
  int index; 
  
  public Index(int index) { 
      this.count = 1; 
      this.index = index; 
  } 
  
  public void incrementCount() { 
      this.count++; 
  } 
} 

public class FirstNonRepeatingCharacter {

  static final int NO_OF_CHARS = 256; 
  
  static HashMap<Character, Index> hm = new HashMap<>(NO_OF_CHARS); 

  static void getCharCountArray(String str) { 
    for (int i = 0; i < str.length();  i++)  {

        if(hm.containsKey(str.charAt(i))) { 
            hm.get(str.charAt(i)).incrementCount(); 
        }
        else { 
            hm.put(str.charAt(i), new Index(i)); 
        } 
              
    }     
  } 

  static int firstNonRepeatingCharacter(String str) {
    getCharCountArray(str); 
    int result = Integer.MAX_VALUE, i; 
    
    for (i = 0; i < str.length();  i++) { 
          // If this character occurs only once and appears 
        // before the current result, then update the result 
        if (hm.get(str.charAt(i)).count == 1 && result > hm.get(str.charAt(i)).index){ 
              result = hm.get(str.charAt(i)).index; 
        } 
            
    }   
        
    return result; 
  }
  public static void main (String[] args) {
    Scanner scanner = new Scanner(System.in);
    System.out.print("Input String: ");
    String str = scanner.nextLine();
    int index = firstNonRepeatingCharacter(str); 
      
    System.out.println(index == Integer.MAX_VALUE ? "Either all characters are repeating " + 
            " or string is empty" : "First non-repeating character is " +  str.charAt(index)); 
  } 
}

File: first_non_repeating_character.c
// C program to find first non-repeating character
#include<stdlib.h>
#include<stdio.h>
#define NO_OF_CHARS 256

/* Returns an array of size 256 containg count
   of characters in the passed char array */
int *getCharCountArray(char *str)
{
   int *count = (int *)calloc(sizeof(int), NO_OF_CHARS);
   int i;
   for (i = 0; *(str+i);  i++)
      count[*(str+i)]++;
   return count;
}

/* The function returns index of first non-repeating
   character in a string. If all characters are repeating
   then returns -1 */
int firstNonRepeating(char *str)
{
  int *count = getCharCountArray(str);
  int index = -1, i;

  for (i = 0; *(str+i);  i++)
  {
    if (count[*(str+i)] == 1)
    {
      index = i;
      break;
    }
  }

  free(count); // To avoid memory leak
  return index;
}

/* Driver program to test above function */
int main()
{
  char str[] = "example to check";
  int index =  firstNonRepeating(str);
  if (index == -1)
    printf("Either all characters are repeating or string is empty");
  else
   printf("First non-repeating character is %c", str[index]);
  getchar();
  return 0;
}
